firebase-crashlytics/src/androidTest/java/com/google/firebase/crashlytics/internal/metadata/RolloutAssignmentListTest.java

```diff
@@ -0,0 +1,137 @@
+          + "\"parameterValue\":\"false\","
+          + "\"templateVersion\":1"
+          + "},"
+          + "{"
+          + "\"rollout_id\":\"rollout_2\""
+          + "\"variantId\":\"control\","
+          + "\"parameterKey\":\"color_feature\","
+          + "\"parameterValue\":\"blue\","
+          + "\"templateVersion\":2"
+          + "}]";
+
+  private static List<RolloutAssignment> ROLLOUTS_STATE_1;
+
+  private static List<RolloutAssignment> ROLLOUTS_STATE_2;
+
+  private static final String ROLLOUT_ASSIGNMENT_JSON =
+      "{"
+          + "\"rolloutId\":\"rollout_1\","
+          + "\"variantId\":\"control\","
+          + "\"parameterKey\":\"my_feature\","
+          + "\"parameterValue\":\"false\","
+          + "\"templateVersion\":1"
+          + "}";
+
+  private static final int MAX_ENTRIES = 64;
+
+  private final RolloutAssignmentList rolloutAssignmentList =
+      new RolloutAssignmentList(MAX_ENTRIES);
+
+  @Override
+  public void setUp() throws Exception {
+    super.setUp();
+
+    ROLLOUTS_STATE_1 = new ArrayList<>();
+    ROLLOUTS_STATE_1.add(
+        RolloutAssignment.create("rollout_1", "my_feature", "false", "control", 1));
+
+    ROLLOUTS_STATE_2 = new ArrayList<>();
+    ROLLOUTS_STATE_2.add(
+        RolloutAssignment.create("rollout_1", "my_feature", "false", "control", 1));
+    ROLLOUTS_STATE_2.add(
+        RolloutAssignment.create("rollout_2", "color_feature", "false", "control", 2));
+  }
+
+  @Test
+  public void testRollAssignmentInit_json() throws JSONException {
+    RolloutAssignment expected =
+        RolloutAssignment.create("rollout_1", "my_feature", "false", "control", 1);
+    RolloutAssignment rolloutAssignmentInitFromJson =
+        RolloutAssignment.create(ROLLOUT_ASSIGNMENT_JSON);
+
+    Truth.assertThat(rolloutAssignmentInitFromJson).isEqualTo(expected);
+  }
+
+  @Test
+  public void testRolloutAssignment_encodeJSON() {
+    RolloutAssignment rolloutAssignment =
+        RolloutAssignment.create("rollout_1", "my_feature", "false", "control", 1);
+    DataEncoder encoder =
+        new JsonDataEncoderBuilder().configureWith(AutoRolloutAssignmentEncoder.CONFIG).build();
+    String data = encoder.encode(rolloutAssignment);
+
+    Truth.assertThat(data).isNotEmpty();
+  }
+
+  @Test
+  public void testUpdateRolloutAssignmentList() throws Exception {
+    final List<String> changed = new ArrayList<String>();
+
+    rolloutAssignmentList.updateMapList(ROLLOUTS_STATE_1);
+
+    new Thread(
+            new Runnable() {
+              @Override
+              public void run() {
+                changed.add("changed");
+                rolloutAssignmentList.updateMapList(ROLLOUTS_STATE_2);
+              }
+            })
+        .start();
+
+    new Thread(
+            new Runnable() {
+              @Override
+              public void run() {
+                List<RolloutAssignment> list = rolloutAssignmentList.getKeysMapList();
+                if (changed.isEmpty()) {
```

Comment:
What is this testing? Is there a race between the two threads? Sometimes changed is empty and sometimes it's not?

Reply:
Yup want to check on the race condition. If you have one thread writing and one thread reading. I think the result is un-predictable. But want to make sure if writing before reading, reading will have the updated value, otherwise reads the previous value